node/core/pvf-checker/src/lib.rs

```diff
@@ -0,0 +1,500 @@
+	/// observed whatsoever. If the leaf is not better then this holds the existing recent block.
+	recent_block: (BlockNumber, Hash),
+	/// The full list of PVFs that are pending pre-checking according to the runtime API. In case
+	/// the API returned an error this list is empty.
+	pending_pvfs: Vec<ValidationCodeHash>,
+}
+
+/// Examines the new leaf and returns the effects of the examination.
+///
+/// Returns `None` if the PVF pre-checking runtime API is not supported for the given leaf hash.
+async fn examine_activation(
+	state: &mut State,
+	sender: &mut impl SubsystemSender,
+	keystore: &SyncCryptoStorePtr,
+	leaf_hash: Hash,
+	leaf_number: BlockNumber,
+) -> Option<ActivationEffect> {
+	let pending_pvfs = match runtime_api::pvfs_require_precheck(sender, leaf_hash).await {
+		Err(runtime_api::RuntimeRequestError::NotSupported) => return None,
+		Err(_) => {
+			tracing::debug!(
+				target: LOG_TARGET,
+				relay_parent = ?leaf_hash,
+				"cannot fetch PVFs that require pre-checking from runtime API",
+			);
+			Vec::new()
+		},
+		Ok(v) => v,
+	};
+
+	let recent_block = match state.recent_block {
+		Some((recent_block_num, recent_block_hash)) if leaf_number < recent_block_num => {
+			// the existing recent block is not worse than the new activation, so leave it.
+			(recent_block_num, recent_block_hash)
+		},
+		_ => (leaf_number, leaf_hash),
+	};
+
+	let new_session_index = match runtime_api::session_index_for_child(sender, leaf_hash).await {
+		Ok(session_index) =>
+			if state.latest_session.map_or(true, |l| l < session_index) {
+				let signing_credentials =
+					check_signing_credentials(sender, keystore, leaf_hash).await;
+				Some((session_index, signing_credentials))
+			} else {
+				None
+			},
+		Err(e) => {
+			tracing::warn!(
+				target: LOG_TARGET,
+				relay_parent = ?leaf_hash,
+				"cannot fetch session index from runtime API: {:?}",
+				e,
+			);
+			None
+		},
+	};
+
+	Some(ActivationEffect { new_session_index, recent_block, pending_pvfs })
+}
+
+/// Checks the active validators for the given leaf. If we have a signing key for one of them,
+/// returns the [`SigningCredentials`].
+async fn check_signing_credentials(
+	sender: &mut impl SubsystemSender,
+	keystore: &SyncCryptoStorePtr,
+	leaf: Hash,
+) -> Option<SigningCredentials> {
+	let validators = match runtime_api::validators(sender, leaf).await {
+		Ok(v) => v,
+		Err(e) => {
+			tracing::warn!(
+				target: LOG_TARGET,
+				?leaf,
+				"error occured during requesting validators: {:?}",
+				e
+			);
+			return None
+		},
+	};
+
+	polkadot_node_subsystem_util::signing_key_and_index(&validators, keystore)
+		.await
+		.map(|(validator_key, validator_index)| SigningCredentials {
+			validator_key,
+			validator_index,
+		})
+}
+
+/// Signs and submits a vote for or against a given validation code.
+///
+/// If the validator already voted for the given code, this function does nothing.
+async fn cast_vote(
```

Comment:
nit: :bike:-shedding a more descriptive name to match the description would be nice

Reply:
fair, will try!